doc/threat_model/Untrusted_Capsule_Isolation.md

```diff
@@ -0,0 +1,23 @@
+Untrusted Capsule Isolation
+===========================
+
+## Isolation Mechanism
+
+Untrusted capsules are limited to what they should be able to access within
+Rust's type system without using `unsafe`. That isolation is implemented by
+banning `unsafe` from use in untrusted capsule code. This isolation is
+vulnerable to code that exploits compiler bugs or bugs in `unsafe` code in
+libraries. When a board integrator chooses to use an untrusted capsule, they are
+responsible for auditing the code of the untrusted capsule to detect potentially
+malicious behavior. This relies in part on Rust's resistance to underhanded
+programming techniques (stealthy obfuscation), and is a weaker form of isolation
+than the hardware-backed isolation used to isolate the kernel (and other
+applications) from applications.
+
+## Impact on Kernel API Design
+
+Kernel APIs should be designed to limit the data that untrusted capsules have
+access to. Trusted kernel code should use capabilities as necessary in its API
+to limit the access that untrusted capsule code has. For example, an API that
+would allow an untrusted capsule to access data external to it should require a
+"trusted" capability.
```

Comment:
I don't think that we currently meet this requirement. Technically capsules like virtual_spi have the ability to access all sorts of external data (all radio data for example) but do not require a capability (they instead require a reference to the chip Spi driver). I think we should probably provide a definition for what a capability is in the context of Tock, and elaborate that either a capability or an exclusive reference to some form of privileged object is required to limit such access. Examples would also probably be useful here.

Reply:
I clarified the wording a bit -- does it match your expectation now? I thought it already worked in the way I described.

Reply:
I like the new wording, but I'm still not sure that Tock currently works this way. The `virtual_spi` capsule can access the data passed by any other capsule that uses SPI (i.e. the contents of a radio packet received over SPI or the value of a sensor reading completed over SPI), and the `virtual_spi` capsule does not require a "trusted" capability (which I interpret to mean a `XCapability` struct that requires `unsafe` to create).

Unless all data passed over SPI is not considered external to the `virtual_spi` capsule?

Reply:
Data passed over SPI is not considered external. Clients of the `virtual_spi` capsule would not be surprised that the `virtual_spi` capsule can see the data passed over SPI.

I'm not sure how to express that more precisely. My conceptual model is "well it's kinda obvious from the interface that the data will flow through that component" but it's hard to formalize.